### PR TITLE
enh(secu): add rule for max session duration for 19 04

### PR DESCRIPTION
**Pull request by sc979**
# Pull Request Template

## Description

Add a rule in the administration form to require a session value lesser or equal to 120min.
Update the user's saved session value in the DB if greater to 120min

image

**Fixes** # (secu)

## Type of change

- [x] Patch fixing an issue (non-breaking change)
- [ ] New functionality (non-breaking change)
- [ ] Breaking change (patch or feature) that might cause side effects breaking part of the Software
- [ ] Updating documentation (missing information, typo...)

## Target serie

- [ ] 2.8.x
- [ ] 18.10.x
- [x] 19.04.x
- [ ] 19.10.x (master)

<h2> How this pull request can be tested ? </h2>

Please describe the **procedure** to verify that the goal of the PR is matched. Provide clear instructions so that it can be **correctly tested**.

Any **relevant details** of the configuration to perform the test should be added.

#### Community contributors & Centreon team

- [ ] I followed the **coding style guidelines** provided by Centreon
- [ ] I have commented my code, especially new **classes**, **functions** or any **legacy code** modified. (***docblock***)
- [ ] I have commented my code, especially **hard-to-understand areas** of the PR.
- [ ] I have made corresponding changes to the **documentation**.
- [x] I have **rebased** my development branch on the base branch (master, maintenance).

#### Centreon team only

- [ ] I have made sure that the **unit tests** related to the story are successful.
- [ ] I have made sure that **unit tests cover 80%** of the code written for the story.
- [ ] I have made sure that **acceptance tests** related to the story are successful (**local and CI**)

### CODE REVIEW

File: www/include/Administration/parameters/DB-Func.php [hunk 1]
```diff
@@ -113,6 +113,21 @@ function is_writable_file_if_exist($path = null)
...
 }
 
+/**
+ * rule to check the session duration value chosen by the user
+ * @param int $value
+ * @param int $valueMax
+ *
+ * @return bool
+ */
+function is_duration_valid(int $value = null)
```
Comment (ganoze) [inline www/include/Administration/parameters/DB-Func.php]:
```suggestion
function is_session_duration_valid(int $value = null)
```

File: www/include/Administration/parameters/DB-Func.php [hunk 2]
```diff
@@ -113,6 +113,21 @@ function is_writable_file_if_exist($path = null)
...
+{
+    if ($value > 0 && $value <= SESSION_DURATION_LIMIT) {
```
Comment (ganoze) [inline www/include/Administration/parameters/DB-Func.php]:
```suggestion
    return ($value > 0 && $value <= SESSION_DURATION_LIMIT);
```

File: www/install/php/Update-19.04.5.php [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+    if ($sessionValue > 120) {
+        $pearDB->query(
+            'UPDATE `options` SET `value` = "120"
+            WHERE `key` = "session_expire"'
+        );
+    }
+} catch (\PDOException $e) {
+    $centreonLog->insertLog(
+        2,
+        "UPGRADE : 19.04.5 Unable to modify session fixation value"
```
Comment (ganoze) [inline www/install/php/Update-19.04.5.php]:
```suggestion
        "UPGRADE : 19.04.5 Unable to modify session expiration value"
```

File: www/install/php/Update-19.04.5.php [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+/**
+ * Update session duration value to the max allowed : 120 min
+ */
+try {
+    $stmt = $pearDB->query(
+        'SELECT `value` FROM `options` WHERE `key` = "session_expire"'
+    );
+    $sessionValue = $stmt->fetch();
+
+    if ($sessionValue > 120) {
```
Comment (ganoze) [inline www/install/php/Update-19.04.5.php]: Could you reuse the _SESSION_DURATION_LIMIT_ constant in some way ?

File: www/include/Administration/parameters/DB-Func.php [hunk 5]
```diff
@@ -113,6 +113,21 @@ function is_writable_file_if_exist($path = null)
...
+ * @param int $valueMax
+ *
+ * @return bool
+ */
+function is_session_duration_valid(int $value = null)
+{
+    return ($value > 0 && $value <= SESSION_DURATION_LIMIT);
+        return true;
+    }
+    return false;
```
Comment (ganoze) [inline www/include/Administration/parameters/DB-Func.php]: Please remove these lines following the above changes ;-)

File: www/include/Administration/parameters/general/form.php [hunk 6]
```diff
@@ -79,6 +81,12 @@
 $form->addElement('text', 'oreon_web_path', _("Centreon Web Directory"), $attrsText);
 
 $form->addElement('text', 'session_expire', _("Sessions Expiration Time"), $attrsText2);
+$form->registerRule('is_duration_valid', 'callback', 'is_duration_valid');
```
Comment (ganoze) [inline www/include/Administration/parameters/general/form.php]:
```suggestion
$form->registerRule('is_session_duration_valid', 'callback', 'is_session_duration_valid');
```

File: www/include/Administration/parameters/general/form.php [hunk 7]
```diff
@@ -79,6 +81,12 @@
...
+$form->addRule(
+    'session_expire',
+    _("This value needs to be an integer lesser than") . " " . SESSION_DURATION_LIMIT . " min",
+    'is_duration_valid'
```
Comment (ganoze) [inline www/include/Administration/parameters/general/form.php]:
```suggestion
    'is_session_duration_valid'
```

File: www/install/php/Update-19.04.5.php (hunk 4 above)
Comment (sc979) [reply www/install/php/Update-19.04.5.php]: I can't, as the value won't be available until the restart of the services after the upgrade.
As the value was set to 7200 sec in the 50-centreon.ini, this value will be the default one after the process is acheived.
Then the customer will be able to modify it, and the form will take in account this new value, as the max value.

File: www/include/Administration/parameters/DB-Func.php (hunk 5 above)
Comment (sc979) [reply www/include/Administration/parameters/DB-Func.php]: done

File: www/include/Administration/parameters/general/form.php (hunk 6 above)
Comment (sc979) [reply www/include/Administration/parameters/general/form.php]: done

File: www/include/Administration/parameters/general/form.php (hunk 7 above)
Comment (sc979) [reply www/include/Administration/parameters/general/form.php]: done

File: lang/fr_FR.UTF-8/LC_MESSAGES/messages.po [hunk 8]
```diff
@@ -14798,3 +14798,6 @@ msgstr "Filtrer par groupe de services"
 
 msgid "Chart"
 msgstr "Graphiques"
+
+msgid "This value needs to be an integer lesser than"
+msgstr "Cette valeur doit être inférieur à"
```
Comment (kduret) [inline lang/fr_FR.UTF-8/LC_MESSAGES/messages.po]:
```suggestion
msgstr "Cette valeur doit être inférieure à"
```

**Comment by sc979** [comment]
I droped the commit for the 19.04.6 upgrade.
The upgrade process will be available in the 19.04.5

**Pull request merged**